File: tools/server_observer_cpp/src/observation_session.cpp
#include "observation_session.hpp"
#include <iostream>
#include <thread>
#include <chrono>

static const int MAX_RETRIES = 3;
static const int TIME_TILL_RETRY = 10;

json ObservationPackage::to_json() const {
    json j = {
        {"game_id", game_id},
        {"headers", headers},
        {"cookies", cookies},
        {"proxy", proxy},
        {"auth", auth.to_json()},
        {"client_version", client_version},
        {"game_server_address", game_server_address}
    };
    
    // Convert time_stamps map to JSON
    json ts = json::object();
    for (const auto& [key, value] : time_stamps) {
        ts[std::to_string(key)] = value;
    }
    j["time_stamps"] = ts;
    
    // Convert state_ids map to JSON
    json si = json::object();
    for (const auto& [key, value] : state_ids) {
        si[std::to_string(key)] = value;
    }
    j["state_ids"] = si;
    
    return j;
}

ObservationPackage ObservationPackage::from_json(const json& j) {
    ObservationPackage pkg;
    pkg.game_id = j.value("game_id", 0);
    pkg.headers = j.value("headers", json::object());
    pkg.cookies = j.value("cookies", json::object());
    pkg.proxy = j.value("proxy", json::object());
    pkg.client_version = j.value("client_version", 207);
    pkg.game_server_address = j.value("game_server_address", "");
    
    if (j.contains("auth")) {
        pkg.auth = AuthDetails::from_json(j["auth"]);
    }
    
    // Convert time_stamps from JSON to map
    if (j.contains("time_stamps") && j["time_stamps"].is_object()) {
        for (const auto& [key, value] : j["time_stamps"].items()) {
            try {
                int k = std::stoi(key);
                pkg.time_stamps[k] = value.get<int>();
            } catch (...) {}
        }
    }
    
    // Convert state_ids from JSON to map
    if (j.contains("state_ids") && j["state_ids"].is_object()) {
        for (const auto& [key, value] : j["state_ids"].items()) {
            try {
                int k = std::stoi(key);
                pkg.state_ids[k] = value.get<std::string>();
            } catch (...) {}
        }
    }
    
    return pkg;
}

ObservationSession::ObservationSession(int game_id,
                                     std::shared_ptr<Account> account,
                                     std::shared_ptr<StaticMapCache> map_cache,
                                     const std::string& storage_path,
                                     const std::string& metadata_path,
                                     const std::string& long_term_storage_path,
                                     int file_size_threshold)
    : game_id(game_id)
    , account(account)
    , map_cache_(map_cache)
    , storage_path_(storage_path)
    , metadata_path_(metadata_path)
    , long_term_storage_path_(long_term_storage_path)
    , file_size_threshold_(file_size_threshold)
    , storage_(nullptr)
{
    next_update_at = std::chrono::system_clock::now();
}

bool ObservationSession::needs_update(std::chrono::system_clock::time_point now) const {
    return now >= next_update_at;
}

std::unique_ptr<ObservationWorker> ObservationSession::create_worker() {
    return std::make_unique<ObservationWorker>(
        account,
        storage_path_,
        game_id,
        package_,
        map_cache_,
        metadata_path_,
        long_term_storage_path_,
        file_size_threshold_
    );
}

void ObservationSession::update_package(const ObservationPackage& other) {
    package_ = other;
}

void ObservationSession::reset() {
    package_ = ObservationPackage();
    ensure_storage()->update_resume_metadata(json::object());
}

RecordingStorage* ObservationSession::ensure_storage() {
    if (!storage_) {
        storage_ = std::make_unique<RecordingStorage>(
            storage_path_,
            metadata_path_,
            long_term_storage_path_,
            file_size_threshold_
        );
    }
    return storage_.get();
}

ObservationWorker::ObservationWorker(std::shared_ptr<Account> account,
                                   const std::string& storage_path,
                                   int game_id,
                                   const ObservationPackage& package,
                                   std::shared_ptr<StaticMapCache> map_cache,
                                   const std::string& metadata_path,
                                   const std::string& long_term_storage_path,
                                   int file_size_threshold)
    : account_(account)
    , game_id_(game_id)
    , package_(package)
    , map_cache_(map_cache)
{
    storage_ = std::make_unique<RecordingStorage>(
        storage_path,
        metadata_path,
        long_term_storage_path,
        file_size_threshold
    );
    storage_->setup_logging();
}

ObservationWorker::~ObservationWorker() {
    if (storage_) {
        storage_->teardown_logging();
    }
}

void ObservationWorker::on_request_response(const json& response) {
    storage_->update_resume_metadata(package_.to_json());
    storage_->save_response(response);
}

bool ObservationWorker::ensure_observation_package() {
    if (package_.game_id != 0) {
        return true;
    }
    
    json resume_metadata = storage_->get_resume_metadata();
    if (!resume_metadata.empty() && resume_metadata.contains("auth")) {
        std::cout << "Resuming from Storage for game " << game_id_ << std::endl;
        package_ = ObservationPackage::from_json(resume_metadata);
        return true;
    }
    
    std::cout << "Observation package not yet created, building package for game " 
             << game_id_ << std::endl;
    package_ = create_observation_package();
    return package_.game_id != 0;
}

ObservationPackage ObservationWorker::create_observation_package() {
    auto hub_itf = account_->get_interface();
    if (!hub_itf) {
        return ObservationPackage();
    }
    
    // Get authentication details
    AuthDetails auth = hub_itf->get_auth_details();
    json headers = hub_itf->get_headers();
    json cookies = hub_itf->get_cookies();
    
    // Build proxy dict
    json proxy = json::object();
    if (!hub_itf->get_proxy_http().empty()) {
        proxy["http"] = hub_itf->get_proxy_http();
    }
    if (!hub_itf->get_proxy_https().empty()) {
        proxy["https"] = hub_itf->get_proxy_https();
    }
    
    // For now, we need to call the Python GameApi to get game server address
    // This is a simplified version - in production you'd need to implement
    // the full game loading logic
    ObservationPackage pkg;
    pkg.game_id = game_id_;
    pkg.headers = headers;
    pkg.cookies = cookies;
    pkg.proxy = proxy;
    pkg.auth = auth;
    pkg.client_version = 207;
    
    // Try to determine game server address
    // This would need proper implementation via Python GameApi
    pkg.game_server_address = "";
    
    return pkg;
}

void ObservationWorker::reset_package() {
    account_->reset_interface();
    package_ = create_observation_package();
}

bool ObservationWorker::ensure_static_map_data(ObservationApi& api, int map_id) {
    if (map_cache_->is_cached(map_id)) {
        return true;
    }
    
    try {
        json static_map_data = api.get_static_map_data(map_id);
        map_cache_->save(map_id, static_map_data);
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Failed to get static map data: " << e.what() << std::endl;
        return false;
    }
}

bool ObservationWorker::is_game_ended(const json& response) {
    if (!response.is_object()) {
        return false;
    }
    
    if (!response.contains("result") || !response["result"].is_object()) {
        return false;
    }
    
    const auto& result = response["result"];
    if (!result.contains("states") || !result["states"].is_object()) {
        return false;
    }
    
    const auto& states = result["states"];
    for (const auto& [key, state] : states.items()) {
        if (state.is_object() && state.contains("gameEnded") && 
            state["gameEnded"].is_boolean() && state["gameEnded"].get<bool>()) {
            return true;
        }
    }
    
    return false;
}

bool ObservationWorker::run() {
    int attempt = 1;
    
    while (true) {
        std::cout << "Starting update for game " << game_id_ << std::endl;
        
        if (!ensure_observation_package()) {
            std::cerr << "Failed to create observation package" << std::endl;
            return false;
        }
        
        try {
            // Create observation API
            ObservationApi api(
                package_.headers,
                package_.cookies,
                package_.proxy,
                package_.auth,
                game_id_,
                package_.game_server_address,
                package_.client_version
            );
            
            // Fetch game state
            json game_state = api.request_game_state(package_.state_ids, package_.time_stamps);
            
            // Update package with new auth and connection details
            package_.auth = api.get_auth();
            package_.cookies = api.get_cookies();
            package_.headers = api.get_headers();
            package_.game_server_address = api.get_game_server_address();
            
            // Save response
            on_request_response(game_state);
            
            // Process map data
            try {
                if (game_state.contains("result") && game_state["result"].is_object()) {
                    const auto& result = game_state["result"];
                    if (result.contains("states") && result["states"].is_object()) {
                        const auto& states = result["states"];
                        if (states.contains("3") && states["3"].is_object()) {
                            const auto& state3 = states["3"];
                            if (state3.contains("map") && state3["map"].is_object()) {
                                const auto& map = state3["map"];
                                if (map.contains("mapID")) {
                                    int map_id = map["mapID"].get<int>();
                                    ensure_static_map_data(api, map_id);
                                }
                            }
                        }
                    }
                }
            } catch (const std::exception& e) {
                // Map data not available or invalid
            }
            
            // Check if game ended
            if (is_game_ended(game_state)) {
                return false;
            }
            
            return true;
            
        } catch (const std::runtime_error& e) {
            std::string error = e.what();
            
            // Handle authentication failure
            if (error.find("Authentication failed") != std::string::npos) {
                if (attempt >= MAX_RETRIES) {
                    std::cerr << "Authentication failed after " << MAX_RETRIES 
                             << " retries" << std::endl;
                    return false;
                }
                
                std::cerr << "Authentication failed, resetting package and retrying..." 
                         << std::endl;
                reset_package();
                attempt++;
                continue;
            }
            
            // Handle server errors
            if (error.find("HTTP status: 5") != std::string::npos) {
                std::cerr << "GameServer returned error, retrying in " 
                         << TIME_TILL_RETRY << " seconds..." << std::endl;
                std::this_thread::sleep_for(std::chrono::seconds(TIME_TILL_RETRY));
                continue;
            }
            
            // Handle network errors
            if (error.find("failed") != std::string::npos || 
                error.find("timeout") != std::string::npos) {
                std::cerr << "GameServer is not responding, retrying in " 
                         << TIME_TILL_RETRY << " seconds..." << std::endl;
                std::this_thread::sleep_for(std::chrono::seconds(TIME_TILL_RETRY));
                continue;
            }
            
            // Unknown error
            throw;
        }
    }
}

File: services/server_observer/src/observation_session.rs
use crate::account_pool::{Account, ProxyConfig};
use crate::hub_interface_wrapper::HubInterfaceWrapper;
use crate::observation_api::{GameServerError, GameServerResult, ObservationApi, ObservationApiError};
use crate::observation_package::ObservationPackage;
use crate::recording_storage::{RecordingStorage, RecordingStorageError};
use crate::redis_publisher::RedisPublisher;
use crate::response_metadata::ResponseMetadata;
use crate::static_map_cache::{StaticMapCache, StaticMapError};
use serde_json;
use std::io;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use thiserror::Error;
use zstd::stream::encode_all;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ObservationError {
    Success = 0,
    GameEnded,
    AuthFailed,
    ServerError,
    NetworkError,
    PackageCreationFailed,
    UnknownError,
}

#[derive(Debug, Clone)]
pub struct ObservationResult {
    pub error_code: ObservationError,
    pub error_message: String,
    pub game_ended: bool,
}

impl ObservationResult {
    pub fn make_success(game_ended: bool) -> Self {
        Self {
            error_code: ObservationError::Success,
            error_message: String::new(),
            game_ended,
        }
    }

    pub fn make_game_ended() -> Self {
        Self {
            error_code: ObservationError::GameEnded,
            error_message: "Game has ended".to_string(),
            game_ended: true,
        }
    }

    pub fn make_auth_failed(game_ended: bool, msg: impl Into<String>) -> Self {
        Self {
            error_code: ObservationError::AuthFailed,
            error_message: msg.into(),
            game_ended,
        }
    }

    pub fn make_server_error(msg: impl Into<String>) -> Self {
        Self {
            error_code: ObservationError::ServerError,
            error_message: msg.into(),
            game_ended: false,
        }
    }

    pub fn make_network_error(game_ended: bool, msg: impl Into<String>) -> Self {
        Self {
            error_code: ObservationError::NetworkError,
            error_message: msg.into(),
            game_ended,
        }
    }

    pub fn make_package_failed(msg: impl Into<String>) -> Self {
        Self {
            error_code: ObservationError::PackageCreationFailed,
            error_message: msg.into(),
            game_ended: false,
        }
    }

    pub fn make_unknown_error(msg: impl Into<String>) -> Self {
        Self {
            error_code: ObservationError::UnknownError,
            error_message: msg.into(),
            game_ended: false,
        }
    }
}

#[derive(Debug, Error)]
pub enum ObservationSessionError {
    #[error("hub interface error: {0}")]
    Hub(#[from] crate::hub_interface_wrapper::HubInterfaceError),
    #[error("observation api error: {0}")]
    Api(#[from] ObservationApiError),
    #[error("recording storage error: {0}")]
    Storage(#[from] RecordingStorageError),
    #[error("static map error: {0}")]
    StaticMap(#[from] StaticMapError),
    #[error("compression error: {0}")]
    Compression(#[from] io::Error),
    #[error("json error: {0}")]
    Json(#[from] serde_json::Error),
}

pub struct ObservationSession {
    pub game_id: i32,
    pub account: Account,
    pub next_update_at: SystemTime,
    pub update_sequence_number: i64,

    map_cache: Option<StaticMapCache>,
    api: Option<ObservationApi>,
    storage_path: String,
    metadata_path: String,
    long_term_storage_path: String,
    file_size_threshold: i64,
    package: ObservationPackage,
    storage: Option<RecordingStorage>,
    attempt: i32,
    redis_publisher: Option<RedisPublisher>,
    hub_interface: Option<HubInterfaceWrapper>,
}

impl ObservationSession {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        game_id: i32,
        account: Account,
        map_cache: Option<StaticMapCache>,
        storage_path: String,
        metadata_path: String,
        long_term_storage_path: String,
        file_size_threshold: i64,
        redis_publisher: Option<RedisPublisher>,
    ) -> Self {
        Self {
            game_id,
            account,
            next_update_at: SystemTime::now(),
            update_sequence_number: 0,
            map_cache,
            api: None,
            storage_path,
            metadata_path,
            long_term_storage_path,
            file_size_threshold,
            package: ObservationPackage::default(),
            storage: None,
            attempt: 1,
            redis_publisher,
            hub_interface: None,
        }
    }

    pub fn set_proxy(&mut self, proxy_config: ProxyConfig) {
        self.account.proxy_config = proxy_config.clone();
        self.package.proxy = proxy_config.clone();
        if let Some(api) = &mut self.api {
            if let Err(err) = api.set_proxy(proxy_config) {
                tracing::warn!(?err, game_id = self.game_id, "failed to apply new proxy to api");
            }
        }
    }

    pub fn increment_attempt(&mut self) {
        self.attempt += 1;
    }

    pub fn reset_attempt(&mut self) {
        self.attempt = 1;
    }

    pub fn get_attempt(&self) -> i32 {
        self.attempt
    }

    fn ensure_storage(&mut self) -> Result<&mut RecordingStorage, ObservationSessionError> {
        if self.storage.is_none() {
            let metadata = if self.metadata_path.is_empty() {
                None
            } else {
                Some(self.metadata_path.as_str())
            };
            let long_term = if self.long_term_storage_path.is_empty() {
                None
            } else {
                Some(self.long_term_storage_path.as_str())
            };
            self.storage = Some(RecordingStorage::new(
                self.storage_path.as_str(),
                metadata,
                long_term,
                self.file_size_threshold,
            )?);
        }
        Ok(self.storage.as_mut().expect("storage must be initialized"))
    }

    fn ensure_observation_package(&mut self) -> bool {
        if self.package.game_id != 0 {
            return true;
        }

        let resume_metadata = match self.ensure_storage() {
            Ok(storage) if storage.has_resume_metadata() => storage.get_resume_metadata(),
            Ok(_) => serde_json::Value::Null,
            Err(err) => {
                tracing::warn!(?err, game_id = self.game_id, "failed to load resume metadata");
                serde_json::Value::Null
            }
        };
        if resume_metadata.is_object() {
            self.package = ObservationPackage::from_json(resume_metadata);
            match ObservationApi::new(
                self.package.headers.clone(),
                self.package.cookies.clone(),
                self.account.proxy_config.clone(),
                self.package.auth.clone(),
                self.game_id,
                self.package.game_server_address.clone(),
                self.package.client_version,
            ) {
                Ok(api) => {
                    self.api = Some(api);
                    return true;
                }
                Err(err) => {
                    tracing::warn!(
                        ?err,
                        game_id = self.game_id,
                        "failed restoring api from resume metadata"
                    );
                }
            }
        }

        self.package = self.create_observation_package();
        self.package.game_id != 0
    }

    fn create_observation_package(&mut self) -> ObservationPackage {
        let proxy_url = if self.account.proxy_config.enabled {
            self.account.proxy_config.to_url()
        } else {
            String::new()
        };

        if self.hub_interface.is_none() {
            match HubInterfaceWrapper::new(proxy_url.clone(), proxy_url) {
                Ok(wrapper) => self.hub_interface = Some(wrapper),
                Err(err) => {
                    tracing::error!(?err, game_id = self.game_id, "failed creating HubInterfaceWrapper");
                    return ObservationPackage::default();
                }
            }
        }

        let Some(hub_itf) = &mut self.hub_interface else {
            return ObservationPackage::default();
        };

        if !hub_itf.is_authenticated() {
            match hub_itf.login(&self.account.username, &self.account.password) {
                Ok(true) => {}
                Ok(false) => {
                    tracing::error!(
                        game_id = self.game_id,
                        account = self.account.username,
                        "hub login returned false"
                    );
                    return ObservationPackage::default();
                }
                Err(err) => {
                    tracing::error!(?err, game_id = self.game_id, "hub login failed");
                    return ObservationPackage::default();
                }
            }
        }

        let game_data = match hub_itf.join_game_as_guest(self.game_id) {
            Ok(data) => data,
            Err(err) => {
                tracing::error!(?err, game_id = self.game_id, "failed to join game as guest");
                return ObservationPackage::default();
            }
        };

        let pkg = ObservationPackage {
            game_id: self.game_id,
            headers: game_data.headers,
            cookies: game_data.cookies,
            proxy: self.account.proxy_config.clone(),
            auth: game_data.auth,
            client_version: game_data.client_version,
            game_server_address: game_data.game_server_address,
            time_stamps: Default::default(),
            state_ids: Default::default(),
        };

        match ObservationApi::new(
            pkg.headers.clone(),
            pkg.cookies.clone(),
            self.account.proxy_config.clone(),
            pkg.auth.clone(),
            self.game_id,
            pkg.game_server_address.clone(),
            pkg.client_version,
        ) {
            Ok(api) => {
                self.api = Some(api);
                pkg
            }
            Err(err) => {
                tracing::error!(?err, game_id = self.game_id, "failed creating ObservationApi");
                ObservationPackage::default()
            }
        }
    }

    pub fn reset_package(&mut self) {
        if let Ok(storage) = self.ensure_storage() {
            let _ = storage.flush_metadata();
        }
        self.hub_interface = None;
        self.package = self.create_observation_package();
    }

    async fn ensure_static_map_data(&mut self, map_id: &str) -> bool {
        let Some(map_cache) = &self.map_cache else {
            return false;
        };
        if map_cache.is_cached(map_id) {
            return true;
        }

        let Some(api) = &self.api else {
            return false;
        };
        match api.get_static_map_data(map_id).await {
            Ok(static_map_data) => map_cache.save(map_id, &static_map_data).await.is_ok(),
            Err(err) => {
                tracing::warn!(?err, game_id = self.game_id, map_id = map_id, "failed static map fetch");
                false
            }
        }
    }

    fn handle_game_server_error(&mut self, result: &GameServerResult) -> ObservationResult {
        match result.error_code {
            GameServerError::AuthError => {
                self.reset_package();
                ObservationResult::make_auth_failed(false, result.error_message.clone())
            }
            GameServerError::HttpError => {
                self.reset_package();
                ObservationResult::make_server_error(result.error_message.clone())
            }
            GameServerError::NetworkError => {
                self.reset_package();
                ObservationResult::make_network_error(false, result.error_message.clone())
            }
            GameServerError::ClientVersionMismatch
            | GameServerError::ServerSwitch
            | GameServerError::ParseError => {
                ObservationResult::make_unknown_error(result.error_message.clone())
            }
            GameServerError::UnknownError | GameServerError::Success => {
                self.reset_package();
                ObservationResult::make_unknown_error(result.error_message.clone())
            }
        }
    }

    fn process_successful_response(
        &mut self,
        result: &GameServerResult,
    ) -> Result<(), ObservationSessionError> {
        if let Some(api) = &self.api {
            api.update_package(&mut self.package);
        }

        let compressed_response = encode_all(result.raw_response.as_bytes(), 3)?;
        let timestamp_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or(Duration::from_secs(0))
            .as_millis() as i64;

        let metadata = ResponseMetadata {
            timestamp: timestamp_ms,
            game_id: self.game_id as i64,
            // Guest observer sessions currently use player_id = 0; this is kept for
            // compatibility while allowing future expansion.
            player_id: 0,
            client_version: self.package.client_version as i64,
            map_id: result.map_id.clone(),
        };

        if let Some(redis) = &self.redis_publisher {
            if let Err(err) = redis.publish_compressed_response(
                &metadata,
                &compressed_response,
            ) {
                tracing::warn!(?err, game_id = self.game_id, "failed redis publish");
            }
        }

        // For on-disk recording, store metadata and raw JSON response together in a
        // single zstd-compressed frame with the following layout:
        //   [4 bytes BE metadata_len][metadata JSON bytes]
        //   [4 bytes BE response_len][raw response JSON bytes]
        let metadata_json = serde_json::to_vec(&metadata)?;
        let response_bytes = result.raw_response.as_bytes();

        let meta_len = metadata_json.len() as u32;
        let resp_len = response_bytes.len() as u32;

        let mut combined = Vec::with_capacity(
            4 + metadata_json.len() + 4 + response_bytes.len(),
        );
        combined.extend_from_slice(&meta_len.to_be_bytes());
        combined.extend_from_slice(&metadata_json);
        combined.extend_from_slice(&resp_len.to_be_bytes());
        combined.extend_from_slice(response_bytes);

        let compressed_recording = encode_all(&combined[..], 3)?;

        let pkg_json = self.package.to_json();
        let storage = self.ensure_storage()?;
        storage.update_resume_metadata(pkg_json);
        storage.save_response(compressed_recording)?;
        Ok(())
    }

    pub async fn run_update_async(&mut self) -> ObservationResult {
        tracing::info!(game_id = self.game_id, attempt = self.attempt, "starting update");

        if let Ok(storage) = self.ensure_storage() {
            let _ = storage.setup_logging();
        }

        let outcome = async {
            if !self.ensure_observation_package() {
                return ObservationResult::make_package_failed("Failed to create observation package");
            }

            let Some(api) = &mut self.api else {
                return ObservationResult::make_package_failed("Observation API not initialized");
            };

            let response = api
                .request_game_state_async(&mut self.package.state_ids, &mut self.package.time_stamps)
                .await;
            let result = api.parse_and_validate_response(
                response,
                &mut self.package.state_ids,
                &mut self.package.time_stamps,
            );

            if !result.success() {
                return self.handle_game_server_error(&result);
            }

            if !result.map_id.is_empty() {
                let _ = self.ensure_static_map_data(&result.map_id).await;
            }

            if let Err(err) = self.process_successful_response(&result) {
                return ObservationResult::make_unknown_error(err.to_string());
            }

            if result.game_ended {
                ObservationResult::make_game_ended()
            } else {
                ObservationResult::make_success(false)
            }
        }
        .await;

        if let Ok(storage) = self.ensure_storage() {
            let _ = storage.teardown_logging();
        }

        outcome
    }
}

impl Drop for ObservationSession {
    fn drop(&mut self) {
        if let Some(storage) = &self.storage {
            let _ = storage.flush_metadata();
            let _ = storage.teardown_logging();
        }
    }
}
